src/stan/model/indexing/lvalue.hpp

```diff
@@ -436,16 +974,17 @@ inline void assign(std::vector<T>& x, const cons_index_list<index_uni, L>& idxs,
  * and size of first dimension of value do not match, or any of
  * the recursive tail assignment dimensions do not match.
  */
-template <typename T, typename I, typename L, typename U>
-std::enable_if_t<!std::is_same<I, index_uni>::value> inline assign(
-    std::vector<T>& x, const cons_index_list<I, L>& idxs,
-    const std::vector<U>& y, const char* name = "ANON", int depth = 0) {
+template <typename T, typename Idx1, typename Idx2, typename U,
+          require_all_std_vector_t<T, U>* = nullptr>
+inline void assign(T&& x, const cons_index_list<Idx1, Idx2>& idxs, U&& y,
+                   const char* name = "ANON", int depth = 0) {
   int x_idx_size = rvalue_index_size(idxs.head_, x.size());
-  math::check_size_match("vector[multi,...] assign sizes", "lhs", x_idx_size,
-                         name, y.size());
+  stan::math::check_size_match("vector[multi,...] assign sizes", "lhs",
+                               x_idx_size, name, y.size());
   for (size_t n = 0; n < y.size(); ++n) {
     int i = rvalue_at(n, idxs.head_);
-    math::check_range("vector[multi,...] assign range", name, x.size(), i);
+    stan::math::check_range("vector[multi,...] assign range", name, x.size(),
+                            i);
     assign(x[i - 1], idxs.tail_, y[n], name, depth + 1);
   }
```

Comment:
One thing we could do here is check if `U` is an rvalue and do something like

```cpp
if (std::is_rvalue_reference<U>::value) {
  for (size_t i = 0; i < y.size(); ++i) {
    assign(x[i], nil_index_list(), std::move(y[i]), name, depth + 1);
  }
} else {
  for (size_t i = 0; i < y.size(); ++i) {
    assign(x[i], nil_index_list(), y[i], name, depth + 1);
  }
}
```

Reply:
That would be beneficial when we have like a vector of Eigen matrices

Reply:
Sure, you can do this.